Sources/SwiftDocC/Infrastructure/Topic Graph/AutomaticCuration.swift

```diff
@@ -50,34 +50,59 @@ public struct AutomaticCuration {
         withTraits variantsTraits: Set<DocumentationDataVariantsTrait>,
         context: DocumentationContext
     ) throws -> [TaskGroup] {
-        // Get any default implementation relationships for this symbol
-        //
-        // It's okay to include all variants here because we just use this set to filter
-        // out these values from automatic curation.
-        let defaultImplementationReferences = Set<String>(
-            (node.semantic as? Symbol)?.defaultImplementationsVariants.allValues
-                .lazy
-                .map(\.variant)
-                .flatMap(\.implementations)
-                .compactMap { implementation in
-                    implementation.reference.url?.absoluteString
-                } ?? []
-        )
+        let languagesFilter = Set(variantsTraits.compactMap {
+            $0.interfaceLanguage.map { SourceLanguage(id: $0) }
+        })
         
-        return try context.children(of: node.reference)
-            // Remove any default implementations
-            .filter({ (reference, kind) -> Bool in
-                return !defaultImplementationReferences.contains(reference.absoluteString)
-            })
+        let children: [ResolvedTopicReference]
+        if languagesFilter.isEmpty {
+            // If the caller requests a combined topic section for all source languages, then the results from the `TopicGraph` will be accurate.
+            
+            // Get any default implementation relationships for this symbol
+            //
+            // It's okay to include all variants here because we just use this set to filter
+            // out these values from automatic curation.
+            let defaultImplementationReferences = Set<String>(
+                (node.semantic as? Symbol)?.defaultImplementationsVariants.allValues
+                    .lazy
+                    .map(\.variant)
+                    .flatMap(\.implementations)
+                    .compactMap { implementation in
+                        implementation.reference.url?.absoluteString
+                    } ?? []
+            )
+            
+            children = context.children(of: node.reference)
+                // Remove any default implementations
+                .filter({ reference, _ -> Bool in
+                    return !defaultImplementationReferences.contains(reference.absoluteString)
+                })
+                .map(\.reference)
+        } else {
+            // If the caller requests topic section for a subset of source languages, then the `TopicGraph` is incapable of producing correct results
+            // because it considers all relationships in all languages. Instead we ask the `PathHierarchy` which is source-language-aware.
+            children = context.linkResolver.localResolver.directDescendants(of: node.reference, languagesFilter: languagesFilter)
```

Comment:
If you moved directDescendants to Node, then this could be as simple as:

```
children = node.directDescendants(languagesFilter: languagesFilter)
```

Reply:
Here `node` is a `DocumentationNode` so moving this API to the node would invert the declaration so that the link resolver is passed as an argument instead:

`LinkResolver . method (Node,         Languages)` vs 
`Node         . method (LinkResolver, Languages)`

for example, something like this:

```
children = node.directDescendants(languagesFilter: languagesFilter, using: context.linkResolver.localResolver)
```